nn/module.go

```diff
@@ -59,12 +63,53 @@ func (m *Module) Init(outer IModule) {
 				// Calling Init in a valid Module: struct{*Module} or struct{Module}
 				m.outer = outer
 				m.isTraining = true // training mode by default.
+				m.name = reflect.TypeOf(outer).String()
 			}
 		}
 	}
 	must(m.outer != nil, "GoTorch requires defining modules via embedding a `Module` struct by value")
 }
 
+// Name returns module name
+func (m *Module) Name() string {
+	return m.name
+}
+
+// Outer returns module outer
+func (m *Module) Outer() IModule {
```

Comment:
The function `weightInit` looks a little bit hacky. I am not sure if it is worthy to change the interface `IModule` for it.

Reply:
The Apply method will be added in the next PR.